File: Program's_Contributed_By_Contributors/Implementation_of_DS/DS_Class.js/Queue.js
// enqueue
// dequeue
// isempty
// peek
// size
class Queue {
    #queue;
    #len;
    #front;
    #rear;
    constructor() {
        this.#queue = [];
        this.#front = -1;
        this.#rear = -1;
        this.#len = 0;
    }
    enqueue(value) {
        this.#queue.push(value);
        this.#rear++;
        this.#len++;
        if (this.#queue.length === 1) {
            this.#front = 0;
        }
    }
    dequeue() {
        if (this.#rear >= 0 && this.#queue[this.#front] !== undefined) {
            this.#len--;
            return this.#queue[this.#front++];
        }
        return null;
    }
    size() {
        return this.#len;
    }
    peek() {
        if (this.#len > 0) {
            return this.#queue[this.#front];
        }
        return null;
    }
    queue() {
        if(this.#front >= this.#rear || this.#len === 0){
            return [];
        }
        let res = [];
        for(let i = this.#front; i<=this.#rear;i++){
            res.push(this.#queue[i]);
        }
        return res;
    }
}
let q = new Queue();
q.enqueue(111);
q.enqueue(112);
q.enqueue(113);
console.log("peek after enqueue:");
console.log(q.peek());
q.enqueue(114);
q.dequeue();
q.dequeue();
console.log("size and peek after dequeue:");
console.log(q.size());
console.log(q.peek());
console.log(q.queue());

File: Program's_Contributed_By_Contributors/C++ Programs/static list/queue.cpp
// Static (Array) Implementation of Queue as List

#include <iostream>
using namespace std;
#define MAXSIZE 50


struct node {
    int info;
    int next;
};


class Queue {
    node node[MAXSIZE];
    int avail, rear, front;

    int getNode() {
        if (avail == -1) {
            cout << "Overflow";
            exit(1);
        }
        int p = avail;
        avail = node[avail].next;
        return p;
    }

    void freeNode(int p) {
        node[p].next = avail;
        avail = p;
    }

    void initialOrdering() {
        for (int i = 0; i < MAXSIZE - 1; i++)
            node[i].next = i + 1;
        node[MAXSIZE - 1].next = -1;
    }

public:
    Queue() {
        initialOrdering();
        avail = 0;
        rear = front = -1;
    }

    bool isEmpty() {
        return front == -1;
    }

    void enqueue(int x) {
        int p = getNode();
        node[p].info = x;
        node[p].next = -1;
        if (rear == -1) 
            front = rear = p;
        else {
            node[rear].next = p;
            rear = p;
        }
    }

    void dequeue() {
        if (isEmpty()) {
            cout << "Underflow\n";
            return;
        }
        int p = front; 
        front = node[p].next;

        int item = node[p].info;
        cout << "Deleted item = " << item << endl;
        if (front == -1) 
            rear = -1;
        freeNode(p);
    }

    void display() {
     
        if (isEmpty()) {
            cout << "Empty queue\n";
            return;
        }
        int i = front;
        while (i != -1) {
            cout << node[i].info << ' ';
            i = node[i].next;
        }
        cout << endl;
    }
};

int main() {
    Queue q;
    for (int i = 0; i < 9; i++)
        q.enqueue(i);
 q.display();
    for (int i = 0;i<6 ; i++)
        q.dequeue();
    q.display();
}